src/file_pipe_log/file_mgr.rs

```diff
@@ -0,0 +1,536 @@
+    pub handle: Arc<F::Handle>,
+    pub format: LogFileFormat,
+}
+
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct FileState {
+    pub first_seq: FileSeq,
+    pub total_len: usize,
+}
+
+pub trait FileCollectionMgr<F: FileSystem> {
+    fn len(&self) -> usize;
+    fn get_fd(&self, file_seq: FileSeq) -> Result<Arc<F::Handle>>;
+    fn file_span(&self) -> (FileSeq, FileSeq);
+    fn push(&mut self, file: FileWithFormat<F>) -> FileState;
+}
+
+/// A collection of files for managing active files.
+pub struct ActiveFileCollection<F: FileSystem> {
+    /// Sequence number of the first file.
+    pub first_seq: FileSeq,
+    pub fds: VecDeque<FileWithFormat<F>>,
+}
+
+impl<F: FileSystem> FileCollectionMgr<F> for ActiveFileCollection<F> {
+    #[inline]
+    fn len(&self) -> usize {
+        self.fds.len()
+    }
+
+    #[inline]
+    fn get_fd(&self, file_seq: FileSeq) -> Result<Arc<F::Handle>> {
+        if !(self.first_seq..self.first_seq + self.fds.len() as u64).contains(&file_seq) {
+            return Err(Error::Corruption("file seqno out of range".to_owned()));
+        }
+        Ok(self.fds[(file_seq - self.first_seq) as usize]
+            .handle
+            .clone())
+    }
+
+    #[inline]
+    fn file_span(&self) -> (FileSeq, FileSeq) {
+        if !self.fds.is_empty() {
+            (self.first_seq, self.first_seq + self.fds.len() as u64 - 1)
+        } else {
+            (0, 0)
+        }
+    }
+
+    #[inline]
+    fn push(&mut self, file: FileWithFormat<F>) -> FileState {
+        self.fds.push_back(file);
+        FileState {
+            first_seq: self.first_seq,
+            total_len: self.fds.len(),
+        }
+    }
+}
+
+impl<F: FileSystem> ActiveFileCollection<F> {
+    #[inline]
+    pub fn new(first_seq: FileSeq, fds: VecDeque<FileWithFormat<F>>) -> Self {
+        Self { first_seq, fds }
+    }
+
+    #[inline]
+    pub fn logical_purge(
+        &mut self,
+        file_seq: FileSeq,
+    ) -> (FileState, FileState, VecDeque<FileWithFormat<F>>) {
+        let prev = FileState {
+            first_seq: self.first_seq,
+            total_len: self.fds.len(),
+        };
+        if (self.first_seq..self.first_seq + self.fds.len() as u64).contains(&file_seq) {
+            let purged = file_seq.saturating_sub(self.first_seq);
+            let purged_files = self.fds.drain(..purged as usize).collect();
+            self.first_seq = file_seq;
+            let current = FileState {
+                first_seq: self.first_seq,
+                total_len: self.fds.len(),
+            };
+            return (prev, current, purged_files);
+        }
+        (prev.clone(), prev, VecDeque::<_>::default())
+    }
+}
+
+/// A collection of files for managing stale files.
+///
+/// Stale files contains two parts:
+/// - `Dummy` part, dummy log files named with `.raftlog.dummy` suffix.
+/// - `Expired` part, obsolete log files manually purged by callers, with
+///   `.raftlog` suffix.
```

Comment:
Why two different types? Why not always rename .raftlog to .reserved?

Reply:
As explained, StaleFileCollection has two part of files:
+ dummy part => generated by engine when starting in cold.
+ expired part => when outer user periodically calls `purge_expired_log`, typically called by TiKV, the expired files will be appended into StaleFileCollection.

And for the problems:
1. why we do not `rename` these expired file when purging?
Here, I just wanna delay the `rename` when `Engine.drop()` to avoid the extra timecost on renaming expired files into dummy files.

2. when and how do we convert expired part into dummy part ?
As mentioned above, the converting progress is delayed to `Drop()`. That is, when `Engine.drop()`,  `LogQueue::Append` will convert all remained `expired` logs into `dummy` logs, with `.raftlog.dummy` suffix.

Reply:
> to avoid the extra timecost on renaming expired files into dummy files.

There is no need. The purge is already done in background.